kusto/cloudinfo.go

```diff
@@ -0,0 +1,138 @@
+package kusto
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+)
+
+// abstraction to query metadata and use this information for providing all
+// information needed for connection string builder to provide all the requisite information
+
+const (
+	metadataEndpoint              = "v1/rest/auth/metadata"
+	defaultAuthEnvVarName         = "AadAuthorityUri"
+	defaultKustoClientAppId       = "db662dc1-0cfe-4e1c-a843-19a68e65be58"
+	defaultPublicLoginUrl         = "https://login.microsoftonline.com"
+	defaultRedirectUri            = "https://microsoft/kustoclient"
+	defaultKustoServiceResourceId = "https://kusto.kusto.windows.net"
+	defaultFirstPartyAuthorityUrl = "https://login.microsoftonline.com/f8cdef31-a31e-4b4a-93e4-5f571e91255a"
+	azureADKey                    = "AzureAD"
+)
+
+// retrieved metadata
+type cloudInfo struct {
+	loginEndpoint          string
+	loginMfaRequired       bool
+	kustoClientAppId       string
+	kustoClientRedirectUri string
+	kustoServiceResourceId string
+	firstPartyAuthorityUrl string
+}
+
+var defaultCloudInfo = &cloudInfo{
+	loginEndpoint:          getEnvOrDefault(defaultAuthEnvVarName, defaultPublicLoginUrl),
+	loginMfaRequired:       false,
+	kustoClientAppId:       defaultKustoClientAppId,
+	kustoClientRedirectUri: defaultRedirectUri,
+	kustoServiceResourceId: defaultKustoServiceResourceId,
+	firstPartyAuthorityUrl: defaultFirstPartyAuthorityUrl,
+}
+
+// cache to query it once per instance
+var cloudInfoCache map[string]*cloudInfo
+
+func RetrieveCloudInfoMetadata(kustoUrl string) (*cloudInfo, error) {
```

Comment:
The signature should be:

```go
func RetrieveCloudInfoMetadata(ctx context.Context, kustoURL string) (CloudInfo, error)
```
If something contacts something else, it likely needs a context object.  If for no other reason that in the future you may want to attach telemetry with this.  If you don't do it now and you need it later, you have to write a second call or make a breaking change.

The other changes to the signature will be discussed below.